doc/developer/design/20210713_S3_sources_with_headers.md

```diff
@@ -0,0 +1,135 @@
+-->
+
+S3 and in the future, multi-file, [GCS][], and other sources have multiple
+independent objects that may have their own headers.
+
+For example, imagine an S3 bucket that has historical data encoded as CSV
+formatted files, one file per day. Each CSV file has its own header line, which
+is semantically meaningful but should not be included as data rows.
+
+This feature adds support for decoding multiple CSV objects without garbage from
+headers making it into query results, and lays the groundwork for future file
+types (e.g. Avro OCF, Parquet) that all require their own header lines.
+
+[GCS]: https://cloud.google.com/storage/docs/
+
+## Goals
+
+<!--
+// Enumerate the concrete goals that are in scope for the project.
+-->
+
+Supporting decoding multiple CSV headers within a framework that will function
+for future multi-object sources.
+
+## Non-Goals
+
+<!--
+// Enumerate potential goals that are explicitly out of scope for the project
+// ie. what could we do or what do we want to do in the future - but are not doing now
+-->
+
+The first version of this will only support the CSV format, Avro OCF objects
+should be extendable with the current work, but will be prioritized separately.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+User-facing changes are limited to an expansion of the ability to combine
+existing formats with S3 sources, for example e.g. `FORMAT CSV WITH HEADER`.
+
+When a user specifies a source format that has a header, we will obtain an
+arbitrary file from the source declaration and inspect its header, making it the
+cannonical schema. All files must have exactly the same schema -- if any are
+encountered that do not match, the dataflow will be put into an error state.
+
+Internally, there are only a few changes that need to be made.
+
+* Add some code in purify that can obtain an object according to the source
+  declaration and use it to extract the schema.
+* Modify the messages sent from sources to include a signifier for "beginning of
+  object", which will cause the decode loop to perform schema validation and
+  drop the header line.
+
+### Purification strategy
+
+We will extend the [purification strategy from existing formats][csv-purify]
+into their combination with S3, including using existing syntax.
+
+[csv-purify]: https://github.com/MaterializeInc/materialize/blob/88cf93c3309ca62/src/sql/src/pure.rs#L480-L501
+
+#### Examples
+
+A CSV source containing files with header lines like `id,value` as:
+
+```sql
+CREATE SOURCE example
+FROM S3 DISCOVER OBJECTS USING BUCKET SCAN 'bucket'
+FORMAT CSV WITH HEADER;
+```
+
+will be rewritten via purify to:
+
+```sql
+CREATE SOURCE example (id, value)
+FROM S3 DISCOVER OBJECTS USING USING BUCKET SCAN 'bucket'
+FORMAT CSV WITH HEADER;
+```
```

Comment:
It occurs to me that our existing purification for CSV files is broken. Since purification doesn't output the expected number of columns, replacing the CSV file with one with a different number of columns and then restarting Materialize can result in a corrupt catalog. I think we need to purify to something like:

```sql
CREATE SOURCE example (id, value)
FROM S3 DISCOVER OBJECTS USING USING BUCKET SCAN 'bucket'
FORMAT CSV WITH 4 COLUMNS, HEADER;
```

And that way the shape of the source is fully independent of the data.

It'd be nice if we could infer the number of columns from the `(id, value)` aliases at the top, but alas, SQL is perfectly happy if you specify _less_ aliases than there are columns:

```
postgres=# create view v (a) as select 1 as b, 2 as c;
CREATE VIEW
postgres=# select * from v;
 a | c 
---+---
 1 | 2
(1 row)

postgres=# 
```

Reply:
There's also no reason we shouldn't support

```
CREATE SOURCE example FORMAT CSV
```

and purify that to

```sql
CREATE SOURCE (column1, column2) example FORMAT CSV WITH 2 COLUMNS
```

or however many columns there actually were on the first line of the file.

Reply:
> alas, SQL is perfectly happy if you specify less aliases than there are columns

Ah, is our current behavior a bug? We give errors if the number of columns specified doesn't exactly match the actual number of columns (it doesn't matter the format, but csv is easy to test):

```console
$ cat /opt/blah.csv
foo,bar
```

```sql
materialize=> create materialized source csv (a) from file '/opt/blah.csv' format csv with 2 columns;
CREATE SOURCE
materialize=> select * from csv;
ERROR:  Decode error: Text: CSV error at record number 1: expected 1 columns, got 2.
materialize=> drop source csv;
DROP SOURCE
materialize=> create materialized source csv (a, b, c) from file '/opt/blah.csv' format csv with 2 columns;
CREATE SOURCE
materialize=> select * from csv;
ERROR:  Decode error: Text: CSV error at record number 1: expected 3 columns, got 2.
```

I was relying on the idea that we were doing this intentionally, otherwise I agree that we need to specify this more concretely.

Reply:
```
> materialize=> create materialized source csv (a, b, c) from file '/opt/blah.csv' format csv with 2 columns;
CREATE SOURCE
materialize=> select * from csv;
ERROR:  Decode error: Text: CSV error at record number 1: expected 3 columns, got 2.
```

I mean this is a bug any way you slice it, right? Shouldn't be possible to ask for a 2 column CSV source and end up with a 3 column one... :/

Reply:
Yeah that's fair.

Reply:
But the first example -- where we error on a row that has two columns because we've explicitly named one column -- looks like we are intentionally treating the column names as a complete description of columns. 

If we are treating the column names as correct & complete, then the second issue is just that we aren't actually doing validation at source creation time -- only at row read time, which is a laziness that still suggests we are trying to treat column name lists as canonical.

I'll propose another normalization that will allow us to get in line with defined SQL semantics for name lists in the future.

Reply:
I've added some syntax that allows us to record column names in the catalog:

```sql
FORMAT CSV WITH (HEADER AND)? NAMED COLUMNS (colname, ..)
```

This allows purification to inspect the header and record what it finds, preventing materialized instances that start up with a CSV file that has a different collection of named columns from suddenly having different columns in the dataflow.

We could add an optional `MATCHING` keyword that specifies that the named columns in the file must match the columns in the SQL syntax, but my current intention is to only verify that the counts are correct, the same way that we hand column renaming as part of the source name.

Reply:
Having two separate and incompatible syntaxes for specifying column names in `CREATE SOURCE` seems pretty unfortunate. Can it work like this instead? We extend the existing syntax to allow specifying both `HEADER` and `WITH <n> COLUMNS` in the same source:

```
CREATE SOURCE csv (<col_alias> [, <col_alias> ...])
FROM ...
FORMAT CSV WITH <csv_opt> [, <csv_opt> ...] [DELIMITED BY <delim>]

csv_opt := HEADER | <n> COLUMNS
```

The post-purification semantics are:

  * `HEADER` controls whether the first row of data is skipped or not.
  * `<n> COLUMNS` is verified against the actual number of columns in the data.
  * Columns are named `column1`, `column2`, etc., unless there is an alias for that column name. So having fewer aliases than `<n>` is supported, but having more aliases than `<n>` is not allowed. Column names post-purification are never determined by actually reading the file.

Then it's just a matter of purification. If a source doesn't specify `WITH <n> COLUMNS`, purify needs to inject that option into the source statement. I think we can say that it's the specification of `WITH <n> COLUMNS` that determines whether any "auto detection" happens. If you don't specify that option, we _have_ to read to one CSV to determine the number of columns, and we'll also helpfully record their names as the column alias if you've specified the `HEADER` option and haven't already specified an alias for that column. 

I think that's perfectly backwards compatible with the existing syntax? And also allows for creating CSV sources with headers even if you don't have any of CSV files available just yet, via e.g. `FORMAT CSV WITH HEADER, 5 COLUMNS`.

Reply:
Yeah, my biggest concern with that syntax was allowing users to specify that column aliases must match header column names, but we can just add a `REQUIRE MATCHING COLUMN NAMES` or something to the CSV with options if it gets requested.

Reply:
So @benesch in your proposal, we entirely ignore the names of columns in the header for all purposes? And only care about the number of them?

Reply:
My reading of the proposal is that we only if the column aliases are present and complete, otherwise we still check the headers. 

Since we check the actual headers as part of purification, this only happens once per source (not source instance) and doesn't happen when we're reading from the catalog.

Reply:
> Columns are named column1, column2, etc., unless there is an alias for that column name

My interpretation of this language was that we would ignore the header names if there is no alias

Reply:
Alright the proposal is now pretty clear about this:

* use `WITH HEADER COLUMNS (name, ..)` for verification
* allow aliases for renaming